hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/AlertsEngineImpl.java

```diff
@@ -762,19 +773,21 @@ private void handleAutoResolvedTriggers() {
 
     private int checkMissingStates() {
         int numMatchingEvals = 0;
-        for (MissingState missingState : missingStates) {
+        Iterator<MissingState> itMissingStates = missingStates.iterator();
```

Comment:
It's handled in two ways.
The updated part is synchronized by the missingStates, and the read part is moved to use an iterator.
So, the iterator itself doesn't solve, that's protected via the synchronized, but moving to an iterator, it check on each element, which then is safe on potential modifications.